docs/design/features/bridging-constraints.md

```diff
@@ -0,0 +1,128 @@
+
+Validating instantiated forms is done always so that a `MethodTable` for an invalid type cannot even be created.
+
+From the runtime perspective, it would be desirable to maintain the invariant that `MethodTable` for a generic type instantiation that doesn't meet its own constraints can never be loaded - an attempt to do so would continue throwing an exception. This has implications on how method bodies that may be violating constraints in certain branches can be compiled. Consider:
+
+```csharp
+//
+// This is pseudo-C# just so that I don't have to write IL.
+// Not expected to compile as C# in this shape, but imagine the straightforward lowering to IL.
+//
+interface IInterface { }
+
+struct Constrained<T> where T : IInterface { }
+
+static void Do<T>()
+{
+    if (typeof(T).IsAssignableFrom(typeof(IInterface))
+    {
+        Constrained<T> someLocal = default;
+        try
+        {
+            someLocal.ToString();
+        }
+        catch (GenericException<Constrained<T>>) { }
+    }
+}
+```
+
+The requirement of never having an invalid type instantiation within the runtime type system means that if we're compiling `Do<int>` (where we cannot allow `Constrained<int>` to exist), the code generator or debugger cannot ask questions about all locals (IL locals are scoped to the method, not to a basic block), exception handling regions, or any IL within the `IsAssignableFrom` check. While it may be reasonable to expect the `IsAssignableFrom` check gets optimized out, it would likely not happen for unoptimized code. Locals and EH regions pose additional programs.
+
+It would greatly simplify impact of this if we were to limit the amount of places within a method body that are allowed to break the constraints. The least impactful and potentially sufficiently powerful change would be allowing to break constraints on `call` instruction only.
+
+Special attention is needed to the `if` checks that guard constraint-breaking calls. While it would be possible to take the approach in the above examples (have language compilers generate `if` blocks guarded with e.g. `IsAssignableFrom`, or other APIs depending on the form of the constraint), it seems preferable to delegate this responsibility to the runtime itself.
+
+Instead of the language compiler generating `if (typeof(T).IsAssignableFrom(typeof(IInterface))` checks or similar, we'll define an IL sequence that would be intrinsically recognized as a _constraints bridge_. The bridge would take the form of:
+
+```
+call RuntimeHelpers.CheckConstraintsForMethodCallThatFollows
+br.false ....
+push arg0
+push arg1
+...
+push argN
+call CodeWithMoreStringentConstraints
+```
+
+The sequence must be within the same basic block and there must not be any other call/callvirt between `CheckConstraintsForMethodCallThatFollows` and the more constrained method call.
```

Comment:
```suggestion
The sequence must be within the same basic block and there must not be any other call/callvirt between `CheckConstraintsForMethodCallThatFollows` and the more constrained method call. The more constrained method call must return `void`.
```

Reply:
What is the expected benefit of this restriction?

Reply:
I don't think the more constrained method call returning void is feasible, it will completely shut down the feature for even simple use cases like wanting to check `if (T implements INumber<T>)` and then doing `x + y` for some `T x; T y`.